lib/statsd/instrument/datagram.rb

```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+class StatsD::Instrument::Datagram < ::String
+  class << self
+    def counter(name, value, type, sample_rate: nil, tags: nil)
+      generic(name, value, -'c', sample_rate: sample_rate, tags: tags)
+    end
+
+
+    # Utility function to convert tags to the canonical form.
+    #
+    # - Tags specified as key value pairs will be converted into an array
+    # - Tags are normalized to only use word characters and underscores.
+    #
+    # @param tags [Array<String>, Hash<String, String>, nil] Tags specified in any form.
+    # @return [Array<String>, nil] the list of tags in canonical form.
+    def normalize_tags(tags)
```

Comment:
The way StatsD process tags is also quite a big source of allocations. Do you think it could be the occasion to be more strict about it?

Reply:
I considered it, but I worry about the migration path.

Reply:
It is possible to provide a separate `DatagramBuilder` class that doesn't do normalization, and make the user responsible. I would not want to enable that by default though.